### DISCUSSION TITLE
Question regarding circle loss

### DISCUSSION THREAD

**Discussion by Bloos** [KevinMusgrave/pytorch-metric-learning#610] [Q&A]
Hi!

I've got a question regarding the implementation of the circle loss. It is implemented like this:
```
new_mat[pos_mask_bool] = (
    -self.gamma
    * torch.relu(self.op - anchor_positive.detach())
    * (anchor_positive - self.delta_p)
)
new_mat[neg_mask_bool] = (
    self.gamma
    * torch.relu(anchor_negative.detach() - self.on)
    * (anchor_negative - self.delta_n)
)

logsumexp_pos = lmu.logsumexp(
    new_mat, keep_mask=pos_mask_bool, add_one=False, dim=1
)
logsumexp_neg = lmu.logsumexp(
    new_mat, keep_mask=neg_mask_bool, add_one=False, dim=1
)

losses = self.soft_plus(logsumexp_pos + logsumexp_neg)
```
or rewritten in mathematical Notation:
![image](https://user-images.githubusercontent.com/33551324/231832753-5b732c15-6684-4932-bc99-93d6d38ebbb1.png)

This is not according to the paper wich defines the loss as
![image](https://user-images.githubusercontent.com/33551324/231830842-3c76653b-2ccf-4e8e-91ce-cf0a12b4717d.png)

When looking at how the formula was derived we can see that the brackets should be like this
![image](https://user-images.githubusercontent.com/33551324/231830785-465d8623-5d8c-401e-88c0-f04f31fe12ea.png)

(multiplication of the term inside of the sum instead of adding positive and negative parts up.

Am I missing something?

**Comment by Bloos** [comment]
I think I've found a mistake in my formulas.
![image](https://user-images.githubusercontent.com/33551324/231833187-85322b13-fb94-4cb6-ace1-05230182b744.png)
is not correct. Actually it is: 
![image](https://user-images.githubusercontent.com/33551324/231833522-24678a0c-9a3a-4e25-a476-b341d91e940b.png)
Which is equivalent to
![image](https://user-images.githubusercontent.com/33551324/231833891-525d2d7d-0419-4028-97ea-bd380d738910.png)
and therefore equivalent to the paper. I'll close the discussion.

Sometimes posing a questions helps :D
